File: 백준/Silver/2178. 미로 탐색/미로 탐색.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().split('\n');

class Queue {
    constructor() {
        this.queue = {};
        this.headIndex = 0;
        this.tailIndex = 0;
    }

    enqueue(value) {
        this.queue[this.tailIndex] = value;
        this.tailIndex++;
    }

    dequeue() {
        if (this.getLength() === 0) return null;
        const value = this.queue[this.headIndex];
        delete this.queue[this.headIndex];
        this.headIndex++;
        return value;
    }

    getLength() {
        return this.tailIndex - this.headIndex;
    }
}

const [n, m] = input[0].split(" ").map(Number);
const visited = [];

// 입력 데이터를 2D 배열로 변환
for (let i = 1; i <= n; i++) {
    let line = input[i].split("").map(Number);
    visited.push(line);
}

// BFS 탐색
const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];

let queue = new Queue();
queue.enqueue([0, 0, 1]); // 시작 지점 [x, y, 거리]
visited[0][0] = 2; // 시작 지점을 방문 처리

while (queue.getLength() > 0) {
    const [curX, curY, dist] = queue.dequeue();

    // 목표 지점 도달 시 거리 출력 후 종료
    if (curX === n - 1 && curY === m - 1) {
        console.log(dist);
        break;
    }

    // 4방향 탐색
    for (let i = 0; i < 4; i++) {
        const nx = curX + dx[i];
        const ny = curY + dy[i];

        // 유효한 좌표이고, 방문하지 않았으며, 이동 가능한 곳인 경우
        if (nx >= 0 && nx < n && ny >= 0 && ny < m && visited[nx][ny] === 1) {
            visited[nx][ny] = 2; // 방문 처리
            queue.enqueue([nx, ny, dist + 1]); // 큐에 추가
        }
    }
}

File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
# 최단거리를 구하는거니까 BFS

from collections import deque

N,M = map(int,input().split())
matrix = []
MIN = float('inf')

q = deque()

for _ in range(N):
  matrix.append(list(map(int, input())))

q.append([0,0,1])

dx =[-1,1,0,0]
dy =[0,0,-1,1]

while q:
  curX,curY, distance = q.popleft()

  for i in range(4):
    nx = dx[i] + curX
    ny = dy[i] + curY
    if nx >= 0 and ny >= 0 and nx < N and ny < M and matrix[nx][ny] == 1:
      matrix[nx][ny] += distance
      q.append([nx,ny, distance + 1])

print(matrix[N-1][M-1])






